### PR TITLE
Save significant IMU accel bias changes learned by the EKF

### PR DESCRIPTION
**Pull request by priseborough**
This is a prototype of a feature that enables accel biases learned by the EKF to be saved between flights. Testing to date has been limited to SITL using the default gazebo  and jmavsim multicopter models with accel bias parameters set to non zero values followed by a re-start and a short auto mission with land and disarm to verify the incorrect bias parameters were updated.

This feature would be useful for operation with IMU's that are not thermally calibrated where operation is performed at a significantly different temperature to the original calibration.

Open Questions:

1. Do we need a user controlled method of enabling or disabling the feature and maybe setting the sensitivity. For example with good quality IMU's that are temperature controlled or thermally calibrated, this feature would be unnecessary and could reduce IMU performance .
2. What if a flight experiences clipping  - should we then label the learned biases as invalid? It is common for  the touchdown to cause clipping so clipping when landing is expected would have to be ignored.
3. How well does this work on hardware.

### CODE REVIEW

File: src/modules/sensors/vehicle_imu/VehicleIMU.hpp [hunk 1]
```diff
@@ -145,13 +158,25 @@ class VehicleIMU : public ModuleParams, public px4::ScheduledWorkItem
...
 	perf_counter_t _gyro_generation_gap_perf{perf_alloc(PC_COUNT, MODULE_NAME": gyro data gap")};
 
 	DEFINE_PARAMETERS(
 		(ParamInt<px4::params::IMU_INTEG_RATE>) _param_imu_integ_rate,
 		(ParamInt<px4::params::IMU_GYRO_RATEMAX>) _param_imu_gyro_ratemax
 	)
+
+	bool _accel_cal_available{false};
+
+	struct AcelCal {
```
Comment (dagar) [inline src/modules/sensors/vehicle_imu/VehicleIMU.hpp:173]:
```suggestion
	struct AccelCal {
```

File: src/modules/ekf2/EKF2.cpp [hunk 2]
```diff
@@ -1035,7 +1060,7 @@ void EKF2::PublishSensorBias(const hrt_abstime &timestamp)
 			accel_bias.copyTo(bias.accel_bias);
 			bias.accel_bias_limit = _params->acc_bias_lim;
 			_ekf.getAccelBiasVariance().copyTo(bias.accel_bias_variance);
-			bias.accel_bias_valid = !_ekf.fault_status_flags().bad_acc_bias;
+			bias.accel_bias_valid = !_ekf.fault_status_flags().bad_acc_bias && _acc_cal_available;
```
Comment (dagar) [inline src/modules/ekf2/EKF2.cpp:1063]: Note there's still the high rate `vehicle_acceleration` topic (analogous to `vehicle_angular_velocity`) that grabs the accel bias from the primary estimator and applies it directly (unless invalid).

File: src/modules/sensors/vehicle_imu/VehicleIMU.cpp [hunk 3]
```diff
@@ -638,4 +656,108 @@ void VehicleIMU::PrintStatus()
...
+
+		if (initialised && bias_estimate.longerThan(0.05f)) {
+			const Vector3f accel_cal_orig{_accel_calibration.offset()};
+			Vector3f accel_cal_offset{_accel_calibration.offset()};
+
+			for (int axis_index = 0; axis_index < 3; axis_index++) {
+				accel_cal_offset(axis_index) += bias_estimate(axis_index);
+			}
+
+			if (_accel_calibration.set_offset(accel_cal_offset)) {
```
Comment (dagar) [inline src/modules/sensors/vehicle_imu/VehicleIMU.cpp:739]: The offset is in the sensor frame. For mag calibration I added a helper in the calibration library.

https://github.com/PX4/PX4-Autopilot/blob/03bd5e0f4841debf5e45d0eadbfb28f131eaaeb1/src/lib/sensor_calibration/Magnetometer.hpp#L89-L93

**Comment by dagar** [comment]
I can't help to ask if we should do gyro as well while we're at it?

If it would help with wider testing we could consider getting this in with a `SENS_` parameter to control the bias actually being saved back to parameters. It could have a debug option that prints an info message with what it would have done.

File: src/modules/ekf2/EKF2.cpp [hunk 4]
```diff
@@ -311,13 +311,12 @@ void EKF2::Run()
 
 	bool imu_updated = false;
 	imuSample imu_sample_new {};
-
+	vehicle_imu_s imu;
 	hrt_abstime imu_dt = 0; // for tracking time slip later
+	imu_updated = _vehicle_imu_sub.update(&imu);
```
Comment (dagar) [inline src/modules/ekf2/EKF2.cpp:316]: At the moment we still have to deal with both sensor paths (vehicle_imu and sensor_combined).

I'll try to finish the sensor_combined purge (and multi-EKF replay) in the near future.

File: src/modules/sensors/vehicle_imu/VehicleIMU.hpp (hunk 1 above)
Comment (priseborough) [reply src/modules/sensors/vehicle_imu/VehicleIMU.hpp]: Fixed

File: src/modules/sensors/vehicle_imu/VehicleIMU.cpp (hunk 3 above)
Comment (priseborough) [reply src/modules/sensors/vehicle_imu/VehicleIMU.cpp]: Fixed

File: src/modules/ekf2/EKF2.cpp (hunk 2 above)
Comment (priseborough) [reply src/modules/ekf2/EKF2.cpp:1063]: Yes, the control loops using acceleration should have the benefits of the latest bias estimate.

**Comment by priseborough** [comment]
> I can't help to ask if we should do gyro as well while we're at it?

Once I have this to a state where you are OK with the implementation, I can do the same for the gyro. Depending in the rate of churn in this area of the code (should be low) it might be better to merge this and do the gyro bias save in a separate PR.

File: src/modules/ekf2/EKF2.cpp (hunk 4 above)
Comment (priseborough) [reply src/modules/ekf2/EKF2.cpp:316]: I need to do a SITL test with multi sensor mode disabled.

File: src/modules/sensors/vehicle_imu/VehicleIMU.cpp [hunk 5]
```diff
@@ -736,7 +736,8 @@ void VehicleIMU::AccelCalibrationUpdate()
 				accel_cal_offset(axis_index) += bias_estimate(axis_index);
 			}
 
-			if (_accel_calibration.set_offset(accel_cal_offset)) {
+			// rotate offsets from body to sensor frame before setting
+			if (_accel_calibration.set_offset(_accel_calibration.rotation().transpose()*accel_cal_offset)) {
```
Comment (dagar) [inline src/modules/sensors/vehicle_imu/VehicleIMU.cpp:740]: Scale factor?

Comment (priseborough) [reply src/modules/sensors/vehicle_imu/VehicleIMU.cpp:740]: The EKF doesn't learn that.

**Comment by dagar** [comment]
Continued and completed in https://github.com/PX4/PX4-Autopilot/pull/18399.

**Pull request closed without merging**